Libraries/Animated/src/AnimatedImplementation.js

```diff
@@ -1880,21 +1921,106 @@ var stagger = function(
 };
 
 type Mapping = {[key: string]: Mapping} | AnimatedValue;
+type EventConfig = {
+  listener?: ?Function;
+  useNativeDriver?: bool;
+};
 
-type EventConfig = {listener?: ?Function};
-var event = function(
-  argMapping: Array<?Mapping>,
-  config?: ?EventConfig,
-): () => void {
-  return function(...args): void {
-    var traverse = function(recMapping, recEvt, key) {
+class AnimatedEvent {
+  _argMapping: Array<?Mapping>;
+  _listener: ?Function;
+  __isNative: bool;
+
+  constructor(
+    argMapping: Array<?Mapping>,
+    config?: ?EventConfig
+  ) {
+    this._argMapping = argMapping;
+    this._listener = config && config.listener;
+    this.__isNative = (config && config.useNativeDriver) || false;
+
+    if (__DEV__) {
+      this._validateMapping();
+    }
+  }
+
+  __attach(viewRef, eventName) {
+    invariant(this.__isNative, 'Only native driven events need to be attached.');
+
+    // Find animated values in `argMapping` and create an array representing their
+    // key path inside the `nativeEvent` object. Ex.: ['contentOffset', 'x'].
+    const eventMappings = [];
+
+    const traverse = (value, path: Array<any>) => {
+      if (value instanceof AnimatedValue) {
+        value.__makeNative();
+
+        invariant(
+          path.indexOf('nativeEvent') >= 0,
+          'Native driven events only support animated values contained inside `nativeEvent`.'
+        );
+        const nativeEventPath = path.slice(path.indexOf('nativeEvent') + 1);
+
+        eventMappings.push({
+          nativeEventPath,
+          animatedValueTag: value.__getNativeTag(),
+        });
+        return;
```

Comment:
can you please refactor this to `if { } else { }` it's difficult to notice the return statement here and this type of construct may lead to bugs in the future